contracts/carrot-app/src/handlers/execute.rs

```diff
@@ -154,13 +154,30 @@ fn withdraw(
     // Only the authorized addresses (admin ?) can withdraw
     app.admin.assert_admin(deps.as_ref(), &info.sender)?;
 
+    let position = get_osmosis_position(deps.as_ref())?;
+    let position_details = position.position.clone().unwrap();
+    // Get app's user and set up authz.
+    let user = get_user(deps.as_ref(), &app)?;
+    let authz = app.auth_z(deps.as_ref(), Some(user.clone()))?;
+
+    // Collect all rewards/incentives if they exist
+    let (collect_rewards_msgs, _) = _inner_claim_rewards(
+        &env,
+        position.clone(),
+        position_details.clone(),
+        user.clone(),
+        authz.clone(),
+    )?;
```

Comment:
Are you sure that reward claim doesn't happen automatically? Even if you withdraw position completely?

Reply:
Wonder if it will fail if there are no rewards and we still call claim rewards? Even if it won't fail I would prefer to not use gas for claiming empty reward

Reply:
This would solve the problem then right ?
```rust
let mut app_response = app
        .response("withdraw")
        .add_attribute("withdraw_amount", withdraw_amount)
        .add_attribute("total_amount", total_amount)
        .add_message(withdraw_msg);

    // Add the collect_rewards_msgs only if rewards are superior to 0
    if !rewards.is_empty() {
        app_response = app_response.add_messages(collect_rewards_msgs);
    }

    Ok(app_response)
```

Reply:
Yeah, but also no need to withdraw if it withdraws all

Reply:
Oh yeah makes sense then 
```rust
 // Withdraw funds
    let (withdraw_msg, withdraw_amount, total_amount, _withdrawn_funds) =
        _inner_withdraw(&env, amount, position, position_details, user, authz)?;

    let mut app_response = app
        .response("withdraw")
        .add_attribute("withdraw_amount", withdraw_amount)
        .add_attribute("total_amount", total_amount)
        .add_message(withdraw_msg);

    // Add the collect_rewards_msgs only if there are rewards OR if we are doing a partial withdraw
    // Context: While partial position withdraws on osmosis keep the rewards unclaimed, full withdraws automatically withdraw rewards 
    if !rewards.is_empty() || amount.is_some() {
        app_response = app_response.add_messages(collect_rewards_msgs);
    }

    Ok(app_response)
```

Reply:
yeah makes sense, then we need something like this
```rust
// Withdraw funds
    let (withdraw_msg, withdraw_amount, total_amount, _withdrawn_funds) =
        _inner_withdraw(&env, amount, position, position_details, user, authz)?;

    let mut app_response = app
        .response("withdraw")
        .add_attribute("withdraw_amount", withdraw_amount)
        .add_attribute("total_amount", total_amount)
        .add_message(withdraw_msg);

    // Add the collect_rewards_msgs only if there are rewards AND if we are doing a partial withdraw
    // Context: While partial position withdraws on osmosis keep the rewards unclaimed, full withdraws automatically withdraw rewards 
    if !rewards.is_empty() && amount.is_some() {
        app_response = app_response.add_messages(collect_rewards_msgs);
    }

    Ok(app_response)
```

Reply:
Yes, it would work for the most cases! Exception when amount equal to position. In theory `_inner_withdraw` can return if it was full withdraw